src/trace/trigger.ts

```diff
@@ -0,0 +1,270 @@
+  return event.requestContext;
+}
+
+function extractCloudFrontRequestEventDistributionId(event: CloudFrontRequestEvent) {
+  return event.Records[0].cf.config.distributionId;
+}
+
+function extractCloudWatchLogsEventDecodedLogs(event: CloudWatchLogsEvent) {
+  const buffer = Buffer.from(event.awslogs.data, "base64");
+  const decompressed = gunzipSync(buffer).toString();
+  return JSON.parse(decompressed);
+}
+
+function extractALBEventARN(event: ALBEvent) {
+  return event.requestContext.elb.targetGroupArn;
+}
+
+function extractCloudWatchEventARN(event: ScheduledEvent) {
+  return event.resources[0];
+}
+
+function extractDynamoDBStreamEventARN(event: DynamoDBStreamEvent) {
+  return event.Records[0].eventSourceARN;
+}
+
+function extractKinesisStreamEventARN(event: KinesisStreamEvent) {
+  return event.Records[0].eventSourceARN;
+}
+
+function extractS3EventARN(event: S3Event) {
+  return event.Records[0].s3.bucket.arn;
+}
+
+function extractSNSEventARN(event: SNSEvent) {
+  return event.Records[0].Sns.TopicArn;
+}
+
+function extractSQSEventARN(event: SQSEvent) {
+  return event.Records[0].eventSourceARN;
+}
+
+/**
+ * parseEventSource parses the triggering event to determine the source
+ * Possible Returns:
+ * api-gateway | application-load-balancer | cloudwatch-logs |
+ * cloudwatch-events | cloudfront | dynamodb | kinesis | s3 | sns | sqs
+ */
+export function parseEventSource(event: any) {
+  let eventSource: string | undefined;
+
+  if (eventType.isAPIGatewayEvent(event) || eventType.isAPIGatewayEventV2(event)) {
+    eventSource = "api-gateway";
+  }
+
+  if (eventType.isALBEvent(event)) {
+    eventSource = "application-load-balancer";
+  }
+
+  if (eventType.isCloudWatchLogsEvent(event)) {
+    eventSource = "cloudwatch-logs";
+  }
+
+  if (eventType.isCloudWatchEvent(event)) {
+    eventSource = "cloudwatch-events";
+  }
+
+  if (eventType.isCloudFrontRequestEvent(event)) {
+    eventSource = "cloudfront";
+  }
+
+  if (eventType.isDynamoDBStreamEvent(event)) {
+    eventSource = "dynamodb";
+  }
+
+  if (eventType.isKinesisStreamEvent(event)) {
+    eventSource = "kinesis";
+  }
+
+  if (eventType.isS3Event(event)) {
+    eventSource = "s3";
+  }
+
+  if (eventType.isSNSEvent(event)) {
+    eventSource = "sns";
+  }
+
+  if (eventType.isSQSEvent(event)) {
+    eventSource = "sqs";
+  }
+  return eventSource;
+}
+
+/**
+ * parseEventSourceARN parses the triggering event to determine the event source's
+ * ARN if available. Otherwise we stitch together the ARN
+ */
+export function parseEventSourceARN(source: string | undefined, event: any, context: Context) {
+  const splitFunctionArn = context.invokedFunctionArn.split(":");
+  const region = splitFunctionArn[3];
+  const accountId = splitFunctionArn[4];
+  const awsARN = getAWSPartitionByRegion(region);
+  let eventSourceARN: string | undefined;
+
+  // e.g. arn:aws:s3:::lambda-xyz123-abc890
+  if (source === "s3") {
+    eventSourceARN = extractS3EventARN(event);
```

Comment:
The event structure may still be malformed, even after we apply our typeguard. We should add a try catch around this method

Reply:
I was curious how you knew that the S3 event structure might be malformed versus others? I wasn't sure if I should try wrapping the other similar methods with a `try...catch`

Reply:
I should have clarified, they could all be malformed, I meant a generic try catch around the entire method.